File: ets2_hashfs/cityhash.py
"""CityHash64 — Google's 2011 hash function as used by SCS HashFS v2.

Exact Python port of the C# implementation from TruckLib (CityHash.cs),
which is itself a port of Google's original cityhash-c.

Test vectors (from TruckLib unit tests):
    cityhash64("käsefondue.txt" UTF-8) = 0x77F93071B1668154
    cityhash64("*")                    = 0x0DAC6B40444905D0
    cityhash64("*def/world/model.tests.sii") = 0x3C6369BC6EFDD668
    cityhash64("")                     = 0x9AE16A3B2F90404F
"""

from __future__ import annotations

import struct

K0: int = 0xC3A5C85C97CB3127
K1: int = 0xB492B66FBE98F273
K2: int = 0x9AE16A3B2F90404F
K3: int = 0xC949D7C7509E6557  # ← different from common K3=0x9DDFEA08EB382D69
MASK64: int = 0xFFFFFFFFFFFFFFFF


def _u32le(data: bytes, off: int = 0) -> int:
    return struct.unpack_from("<I", data, off)[0]


def _u64le(data: bytes, off: int = 0) -> int:
    return struct.unpack_from("<Q", data, off)[0]


def _rot64(val: int, shift: int) -> int:
    if shift == 0:
        return val
    return ((val >> shift) | (val << (64 - shift))) & MASK64


def _rot64_at_least_1(val: int, shift: int) -> int:
    """Rotate right — used in HashLen0To16 for len>8."""
    return ((val >> shift) | (val << (64 - shift))) & MASK64


def _shift_mix(val: int) -> int:
    return (val ^ (val >> 47)) & MASK64


def _hash128to64(lo: int, hi: int) -> int:
    """Hash 128 input bits down to 64 bits. Murmur-inspired."""
    k_mul: int = 0x9DDFEA08EB382D69
    a = ((lo ^ hi) * k_mul) & MASK64
    a ^= a >> 47
    b = ((hi ^ a) * k_mul) & MASK64
    b ^= b >> 47
    return (b * k_mul) & MASK64


def _hash_len16(u: int, v: int) -> int:
    return _hash128to64(u, v)


def _hash_len_0_to_16(data: bytes) -> int:
    L = len(data)
    if L > 8:
        a = _u64le(data, 0)
        b = _u64le(data, L - 8)
        return _hash_len16(a, _rot64_at_least_1(b + L, L)) ^ b
    if L >= 4:
        a = _u32le(data, 0)
        return _hash_len16(L + (a << 3), _u32le(data, L - 4))
    if L > 0:
        a = data[0]
        b = data[L >> 1]
        c = data[L - 1]
        y = a + (b << 8)
        z = L + (c << 2)
        return _shift_mix((y * K2) & MASK64 ^ (z * K3) & MASK64) * K2 & MASK64
    return K2


def _hash_len_17_to_32(data: bytes) -> int:
    L = len(data)
    a = (_u64le(data, 0) * K1) & MASK64
    b = _u64le(data, 8)
    c = (_u64le(data, L - 8) * K2) & MASK64
    d = (_u64le(data, L - 16) * K0) & MASK64
    return _hash_len16(
        (_rot64(a - b, 43) + _rot64(c, 30) + d) & MASK64,
        (a + _rot64(b ^ K3, 20) - c + L) & MASK64,
    )


def _hash_len_33_to_64(data: bytes) -> int:
    L = len(data)
    z = _u64le(data, 24)
    a = (_u64le(data, 0) + (L + _u64le(data, L - 16)) * K0) & MASK64
    b = _rot64(a + z, 52)
    c = _rot64(a, 37)
    a = (a + _u64le(data, 8)) & MASK64
    c = (c + _rot64(a, 7)) & MASK64
    a = (a + _u64le(data, 16)) & MASK64
    vf = a + z
    vs = b + _rot64(a, 31) + c
    a = _u64le(data, 16) + _u64le(data, L - 32)
    z = _u64le(data, L - 8)
    b = _rot64(a + z, 52)
    c = _rot64(a, 37)
    a = (a + _u64le(data, L - 24)) & MASK64
    c = (c + _rot64(a, 7)) & MASK64
    a = (a + _u64le(data, L - 16)) & MASK64
    wf = a + z
    ws = b + _rot64(a, 31) + c
    r = _shift_mix(((vf + ws) * K2 + (wf + vs) * K0) & MASK64)
    return _shift_mix((r * K0 + vs) & MASK64) * K2 & MASK64


def _weak_hash_len_32_with_seeds(w: int, x: int, y: int, z: int, a: int, b: int) -> tuple[int, int]:
    a = (a + w) & MASK64
    b = _rot64(b + a + z, 21)
    c = a
    a = (a + x) & MASK64
    a = (a + y) & MASK64
    b = (b + _rot64(a, 44)) & MASK64
    return (a + z) & MASK64, (b + c) & MASK64


def _weak_hash_len_32_with_seeds_buf(data: bytes, off: int, a: int, b: int) -> tuple[int, int]:
    return _weak_hash_len_32_with_seeds(
        _u64le(data, off),
        _u64le(data, off + 8),
        _u64le(data, off + 16),
        _u64le(data, off + 24),
        a, b,
    )


def cityhash64(data: bytes) -> int:
    """Compute the CityHash64 of a byte string.

    Exact port of the C# implementation from TruckLib (CityHash.cs),
    which is Google's original CityHash64 (2011) used by SCS HashFS v2.
    """
    L = len(data)
    if L <= 16:
        return _hash_len_0_to_16(data)
    if L <= 32:
        return _hash_len_17_to_32(data)
    if L <= 64:
        return _hash_len_33_to_64(data)

    # For strings over 64 bytes: hash the end first, then process 64-byte chunks.
    x = _u64le(data, L - 40)
    y = (_u64le(data, L - 16) + _u64le(data, L - 56)) & MASK64
    z = _hash_len16(_u64le(data, L - 48) + L, _u64le(data, L - 24))
    v1, v2 = _weak_hash_len_32_with_seeds_buf(data, L - 64, L, z)
    w1, w2 = _weak_hash_len_32_with_seeds_buf(data, L - 32, (y + K1) & MASK64, x)
    x = (x * K1 + _u64le(data, 0)) & MASK64

    # Decrease len to nearest multiple of 64.
    remaining = (L - 1) & ~63
    pos = 0
    while remaining > 0:
        x = (_rot64((x + y + v1 + _u64le(data, pos + 8)) & MASK64, 37) * K1) & MASK64
        y = (_rot64((y + v2 + _u64le(data, pos + 48)) & MASK64, 42) * K1) & MASK64
        x ^= w2
        y = (y + v1 + _u64le(data, pos + 40)) & MASK64
        z = (_rot64(z + w1, 33) * K1) & MASK64
        v1, v2 = _weak_hash_len_32_with_seeds_buf(data, pos, (v2 * K1) & MASK64, (x + w1) & MASK64)
        w1, w2 = _weak_hash_len_32_with_seeds_buf(data, pos + 32, (z + w2) & MASK64, (y + _u64le(data, pos + 16)) & MASK64)
        z, x = x, z  # swap
        pos += 64
        remaining -= 64

    return _hash_len16(
        _hash_len16(v1, w1) + (_shift_mix(y) * K1 + z) & MASK64,
        _hash_len16(v2, w2) + x,
    )

File: crates/map-parser/src/cityhash.rs
//! CityHash64 — the variant used by SCS HashFS.
//!
//! Ported 1:1 from `TruckLib.HashFs/CityHash.cs` (sk-zk), which itself is a
//! port of `cityhash-c` by Alexander Nusov, which is a port of Google's
//! original CityHash 1.0.x. **This is intentionally NOT Google's CityHash
//! 1.1.x**, which is what most third-party Rust/Python crates ship — they
//! produce different hashes for non-empty input and do **not** match SCS
//! archive entry hashes.
//!
//! The empirical proof lives in `tests/real_archive_test.rs`: the hash of
//! `"automat"` computed here matches a real entry in ETS2 1.55 `base.scs`,
//! while the previous Google-1.1.x-style implementation matched none of
//! three known top-level paths.
//!
//! ## Verified test vectors (all from TruckLib's `CityHash.CityHash64`)
//! ```text
//! cityhash64(b"")            = 0x9AE16A3B2F90404F  (== K2)
//! cityhash64(b"abc")         = 0x3A912F483A4ECE31
//! cityhash64(b"automat")     = 0x56BC42EECBC73F2F  ← present in base.scs
//! cityhash64(b"manifest.sii")= 0xB97FFF7CE7377C95
//! ```

// Some primes between 2^63 and 2^64 for various uses.
const K0: u64 = 0xc3a5c85c97cb3127;
const K1: u64 = 0xb492b66fbe98f273;
const K2: u64 = 0x9ae16a3b2f90404f;
const K3: u64 = 0xc949d7c7509e6557;

/// Multiplier for `Hash128To64` — Murmur-inspired.
const K_MUL: u64 = 0x9ddfea08eb382d69;

/// Compute CityHash64 (1.0.x variant, as used by SCS HashFS).
pub fn cityhash64(data: &[u8]) -> u64 {
    let len = data.len();
    if len <= 16 {
        return hash_len_0_to_16(data);
    }
    if len <= 32 {
        return hash_len_17_to_32(data);
    }
    if len <= 64 {
        return hash_len_33_to_64(data);
    }
    hash_len_65_plus(data)
}

// ── helpers ──────────────────────────────────────────────────────────

fn fetch32(s: &[u8], pos: usize) -> u64 {
    u32::from_le_bytes(s[pos..pos + 4].try_into().unwrap()) as u64
}

fn fetch64(s: &[u8], pos: usize) -> u64 {
    u64::from_le_bytes(s[pos..pos + 8].try_into().unwrap())
}

/// Bitwise right rotate. `shift` is in bits (0..=63).
fn rotate(val: u64, shift: u32) -> u64 {
    if shift == 0 {
        val
    } else {
        val.rotate_right(shift)
    }
}

/// Right rotate by **at least one** bit. Caller guarantees `shift >= 1`.
fn rotate_by_at_least_1(val: u64, shift: u32) -> u64 {
    val.rotate_right(shift)
}

fn shift_mix(val: u64) -> u64 {
    val ^ (val >> 47)
}

/// Hash 128 input bits down to 64. Murmur-inspired.
fn hash128to64(lo: u64, hi: u64) -> u64 {
    let mut a = (lo ^ hi).wrapping_mul(K_MUL);
    a ^= a >> 47;
    let mut b = (hi ^ a).wrapping_mul(K_MUL);
    b ^= b >> 47;
    b.wrapping_mul(K_MUL)
}

fn hash_len_16(u: u64, v: u64) -> u64 {
    hash128to64(u, v)
}

// ── length-specific sub-hashers ─────────────────────────────────────

fn hash_len_0_to_16(s: &[u8]) -> u64 {
    let len = s.len();
    if len > 8 {
        let a = fetch64(s, 0);
        let b = fetch64(s, len - 8);
        return hash_len_16(
            a,
            rotate_by_at_least_1(b.wrapping_add(len as u64), len as u32),
        ) ^ b;
    }
    if len >= 4 {
        let a = fetch32(s, 0);
        return hash_len_16(
            (len as u64).wrapping_add(a << 3),
            fetch32(s, len - 4),
        );
    }
    if len > 0 {
        let a = s[0] as u64;
        let b = s[len >> 1] as u64;
        let c = s[len - 1] as u64;
        let y = a.wrapping_add(b << 8);
        let z = (len as u64).wrapping_add(c << 2);
        return shift_mix(y.wrapping_mul(K2) ^ z.wrapping_mul(K3)).wrapping_mul(K2);
    }
    K2
}

fn hash_len_17_to_32(s: &[u8]) -> u64 {
    let len = s.len();
    let a = fetch64(s, 0).wrapping_mul(K1);
    let b = fetch64(s, 8);
    let c = fetch64(s, len - 8).wrapping_mul(K2);
    let d = fetch64(s, len - 16).wrapping_mul(K0);
    hash_len_16(
        rotate(a.wrapping_sub(b), 43)
            .wrapping_add(rotate(c, 30))
            .wrapping_add(d),
        a.wrapping_add(rotate(b ^ K3, 20))
            .wrapping_sub(c)
            .wrapping_add(len as u64),
    )
}

fn hash_len_33_to_64(s: &[u8]) -> u64 {
    let len = s.len();
    let mut z = fetch64(s, 24);
    let mut a = fetch64(s, 0).wrapping_add(
        ((len as u64).wrapping_add(fetch64(s, len - 16))).wrapping_mul(K0),
    );
    let mut b = rotate(a.wrapping_add(z), 52);
    let mut c = rotate(a, 37);
    a = a.wrapping_add(fetch64(s, 8));
    c = c.wrapping_add(rotate(a, 7));
    a = a.wrapping_add(fetch64(s, 16));
    let vf = a.wrapping_add(z);
    let vs = b.wrapping_add(rotate(a, 31)).wrapping_add(c);
    a = fetch64(s, 16).wrapping_add(fetch64(s, len - 32));
    z = fetch64(s, len - 8);
    b = rotate(a.wrapping_add(z), 52);
    c = rotate(a, 37);
    a = a.wrapping_add(fetch64(s, len - 24));
    c = c.wrapping_add(rotate(a, 7));
    a = a.wrapping_add(fetch64(s, len - 16));
    let wf = a.wrapping_add(z);
    let ws = b.wrapping_add(rotate(a, 31)).wrapping_add(c);
    let r = shift_mix(
        vf.wrapping_add(ws)
            .wrapping_mul(K2)
            .wrapping_add(wf.wrapping_add(vs).wrapping_mul(K0)),
    );
    shift_mix(r.wrapping_mul(K0).wrapping_add(vs)).wrapping_mul(K2)
}

/// Returns `(first, second)` of the C# `Uint128`.
fn weak_hash_len_32_with_seeds_raw(
    w: u64,
    x: u64,
    y: u64,
    z: u64,
    a0: u64,
    b0: u64,
) -> (u64, u64) {
    let mut a = a0.wrapping_add(w);
    let mut b = rotate(b0.wrapping_add(a).wrapping_add(z), 21);
    let c = a;
    a = a.wrapping_add(x);
    a = a.wrapping_add(y);
    b = b.wrapping_add(rotate(a, 44));
    (a.wrapping_add(z), b.wrapping_add(c))
}

fn weak_hash_len_32_with_seeds(s: &[u8], pos: usize, a: u64, b: u64) -> (u64, u64) {
    weak_hash_len_32_with_seeds_raw(
        fetch64(s, pos),
        fetch64(s, pos + 8),
        fetch64(s, pos + 16),
        fetch64(s, pos + 24),
        a,
        b,
    )
}

fn hash_len_65_plus(s: &[u8]) -> u64 {
    let len = s.len();
    // Hash the end first, then loop with 56 bytes of state (v, w, x, y, z).
    let mut x = fetch64(s, len - 40);
    let mut y = fetch64(s, len - 16).wrapping_add(fetch64(s, len - 56));
    let mut z = hash_len_16(
        fetch64(s, len - 48).wrapping_add(len as u64),
        fetch64(s, len - 24),
    );
    let mut v = weak_hash_len_32_with_seeds(s, len - 64, len as u64, z);
    let mut w = weak_hash_len_32_with_seeds(s, len - 32, y.wrapping_add(K1), x);
    x = x.wrapping_mul(K1).wrapping_add(fetch64(s, 0));

    // Decrease len to nearest multiple of 64 and process 64-byte chunks.
    let mut remaining = (len - 1) & !63usize;
    let mut pos = 0usize;
    loop {
        x = rotate(
            x.wrapping_add(y)
                .wrapping_add(v.0)
                .wrapping_add(fetch64(s, pos + 8)),
            37,
        )
        .wrapping_mul(K1);
        y = rotate(
            y.wrapping_add(v.1).wrapping_add(fetch64(s, pos + 48)),
            42,
        )
        .wrapping_mul(K1);
        x ^= w.1;
        y = y.wrapping_add(v.0).wrapping_add(fetch64(s, pos + 40));
        z = rotate(z.wrapping_add(w.0), 33).wrapping_mul(K1);
        v = weak_hash_len_32_with_seeds(
            s,
            pos,
            v.1.wrapping_mul(K1),
            x.wrapping_add(w.0),
        );
        w = weak_hash_len_32_with_seeds(
            s,
            pos + 32,
            z.wrapping_add(w.1),
            y.wrapping_add(fetch64(s, pos + 16)),
        );
        std::mem::swap(&mut z, &mut x);

        pos += 64;
        remaining = remaining.saturating_sub(64);
        if remaining == 0 {
            break;
        }
    }

    hash_len_16(
        hash_len_16(v.0, w.0)
            .wrapping_add(shift_mix(y).wrapping_mul(K1))
            .wrapping_add(z),
        hash_len_16(v.1, w.1).wrapping_add(x),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Empty input → K2.
    #[test]
    fn empty_string_returns_k2() {
        assert_eq!(cityhash64(b""), 0x9AE16A3B2F90404F);
    }

    /// Verified against TruckLib `CityHash.CityHash64` on 2026-05-09 via dotnet.
    #[test]
    fn trucklib_vector_abc() {
        assert_eq!(cityhash64(b"abc"), 0x3A912F483A4ECE31);
    }

    /// Verified against TruckLib **and** confirmed present in
    /// ETS2 1.55 `base.scs` index1 — the empirical regression sentinel.
    #[test]
    fn trucklib_vector_automat() {
        assert_eq!(cityhash64(b"automat"), 0x56BC42EECBC73F2F);
    }

    /// Verified against TruckLib `CityHash.CityHash64`.
    #[test]
    fn trucklib_vector_manifest_sii() {
        assert_eq!(cityhash64(b"manifest.sii"), 0xB97FFF7CE7377C95);
    }

    /// Verified against TruckLib (string length 11 → 17_to_32 path). Was a
    /// useful boundary check while porting.
    #[test]
    fn trucklib_vector_version_sii() {
        assert_eq!(cityhash64(b"version.sii"), 0x02D338506339918D);
    }

    /// Length 18 → 17_to_32 path, longer ASCII path.
    #[test]
    fn trucklib_vector_road_sii() {
        assert_eq!(
            cityhash64(b"def/world/road.sii"),
            0x97E6A16838335F87,
        );
    }

    /// Length 29 → 17_to_32 path edge.
    #[test]
    fn trucklib_vector_sector_base() {
        assert_eq!(
            cityhash64(b"map/europe/sec+0000+0000.base"),
            0xC6C0364995A4439A,
        );
    }

    #[test]
    fn deterministic() {
        let a = cityhash64(b"def/world/road.sii");
        let b = cityhash64(b"def/world/road.sii");
        assert_eq!(a, b);
    }

    #[test]
    fn different_paths_differ() {
        assert_ne!(cityhash64(b"foo"), cityhash64(b"bar"));
    }
}
